File: pivot/web/live-trade.mjs
// Live trade card (4.5.2): pure render helpers over the live_trade / last_trade payloads.
// The backend computes every number from the app's own trade record and quotes; this module
// only words and draws them. No DOM, no fetch, and nothing here can change a trade or an order.
import {escape as esc} from './model.mjs';

const text=value=>typeof value==='string'?value.trim():'';
const num=value=>{if(value===null||value===undefined||value==='')return null;const n=Number(value);return Number.isFinite(n)?n:null;};
const obj=value=>value&&typeof value==='object'&&!Array.isArray(value)?value:null;
const usd=new Intl.NumberFormat('en-US',{style:'currency',currency:'USD'});
const etTime=new Intl.DateTimeFormat('en-US',{timeZone:'America/New_York',hour:'numeric',minute:'2-digit'});
const etDay=new Intl.DateTimeFormat('en-US',{timeZone:'America/New_York',weekday:'short',month:'short',day:'numeric'});
const etKey=new Intl.DateTimeFormat('en-CA',{timeZone:'America/New_York',year:'numeric',month:'2-digit',day:'2-digit'});
// A live mark older than this reads as delayed on screen (the executor trades only on quotes under 15 s).
export const LIVE_STALE_SECONDS=30;
// How often the page asks for the live view while a trade is open.
export const LIVE_POLL_MS=3000;

export function price(value) {
  const n=num(value);if(n===null||n<=0)return '—';
  // Fractional shares of a ~$35 ETF move in cents; show 2 decimals, 3 for sub-cent averages.
  const digits=Math.abs(n*100-Math.round(n*100))>1e-6?3:2;
  return '$'+n.toLocaleString('en-US',{minimumFractionDigits:2,maximumFractionDigits:digits});
}
export function signedMoney(value) {
  const n=num(value);if(n===null)return '—';
  const rounded=Math.round(n*100)/100;
  return (rounded>0?'+':rounded<0?'−':'')+usd.format(Math.abs(rounded));
}
export function signedPercent(value) {
  const n=num(value);if(n===null)return '—';
  return (n>0?'+':n<0?'−':'')+Math.abs(n).toFixed(2)+'%';
}
function signedR(value) {
  const n=num(value);if(n===null)return '';
  return (n>0?'+':n<0?'−':'')+Math.abs(n).toFixed(2)+'R';
}
export function timeET(at) { const t=Date.parse(at);return Number.isFinite(t)?etTime.format(t)+' ET':'—'; }
function dayET(at,now) {
  const t=Date.parse(at);if(!Number.isFinite(t))return '';
  return etKey.format(t)===etKey.format(now)?'today':etDay.format(t);
}
export function duration(seconds) {
  const s=num(seconds);if(s===null||s<0)return '—';
  if(s<60)return `${Math.floor(s)}s`;
  const minutes=Math.floor(s/60);if(minutes<60)return `${minutes} min`;
  return `${Math.floor(minutes/60)} h ${minutes%60} min`;
}
function shares(value) {
  const n=num(value);if(n===null)return '—';
  return n.toLocaleString('en-US',{maximumFractionDigits:6});
}
export function tone(live) {
  const n=num(obj(live)?.pnl?.dollars);
  return n===null?'flat':n>0.004?'up':n<-0.004?'down':'flat';
}

// Everything the card shows, derived once so tests can check the words without HTML.
export function liveTradeView(live,{now=Date.now(),elapsed=0}={}) {
  live=obj(live);if(!live)return null;
  const entry=obj(live.entry)||{},current=obj(live.current),pnl=obj(live.pnl),stop=obj(live.stop),target=obj(live.target);
  const symbol=text(live.symbol)||'QQQ',long=live.direction!=='short';
  const age=num(current?.age_seconds)===null?null:num(current.age_seconds)+Math.max(0,elapsed);
  const stale=!current||current.source!=='quote'||age===null||age>LIVE_STALE_SECONDS;
  const exit=obj(live.session_exit),left=exit&&num(exit.seconds_left)!==null?Math.max(0,num(exit.seconds_left)-Math.max(0,elapsed)):null;
  const stopLevel=num(stop?.price),targetLevel=num(target?.price),entryPrice=num(entry.price);
  let entryAt=null;
  if(stopLevel!==null&&targetLevel!==null&&entryPrice!==null&&stopLevel!==targetLevel)
    entryAt=Math.min(1,Math.max(0,(long?entryPrice-stopLevel:stopLevel-entryPrice)/(long?targetLevel-stopLevel:stopLevel-targetLevel)));
  const order=obj(stop?.order);
  const view={
    symbol,stage:text(live.stage),stageText:text(live.stage_text)||'Managing',label:text(live.label)||`Socrates ${symbol}`,
    proxy:live.proxy===true,tone:tone(live),
    bought:entryPrice!==null?`Bought ${shares(entry.qty)} ${symbol} at ${price(entry.price)}${entry.at?` · ${timeET(entry.at)}`:''}`:'Waiting for the purchase to fill',
    cost:entry.cost?price(entry.cost):'—',
    now:current?price(current.price):'—',
    nowLabel:current?.source==='quote'?(long?'Now (bid)':'Now (ask)'):'Now (last trade)',
    updated:!current?'Waiting for a live price':current.source==='quote'?(age<=2?'Price just now':`Price ${duration(age)} ago`):'Price from the Alpaca position',
    stale,
    pnl:pnl?signedMoney(pnl.dollars):'—',
    pnlDetail:pnl?[signedPercent(pnl.percent),signedR(pnl.r)].filter(Boolean).join(' · '):'Shown once the purchase has filled',
    value:pnl&&num(entry.cost)!==null&&num(pnl.dollars)!==null?price(num(entry.cost)+num(pnl.dollars)):'—',
    progress:num(live.progress),entryAt,levels:{stop:stopLevel,target:targetLevel,entry:entryPrice},
    stopPrice:price(stop?.price),targetPrice:price(target?.price),
    target:target?{price:price(target.price),result:signedMoney(target.result_if_hit),percent:signedPercent(target.result_percent),
      away:num(target.distance_percent)===null?'':num(target.distance_percent)<=0?'reached':`${Math.abs(num(target.distance_percent)).toFixed(2)}% away`}:null,
    stop:stop?{price:price(stop.price),result:signedMoney(stop.result_if_hit),percent:signedPercent(stop.result_percent),
      away:num(stop.distance_percent)===null?'':num(stop.distance_percent)<=0?'reached':`${Math.abs(num(stop.distance_percent)).toFixed(2)}% away`,
      order:text(order?.text)||'status not confirmed yet',working:order?.working===true}:null,
    close:exit?{at:timeET(exit.at),left:left===null?'':left<=0?'now':`in ${duration(left)}`}:null,
    signal:obj(live.signal),message:text(live.message),exitReason:text(obj(live.exit)?.reason),
    track:Array.isArray(live.track)?live.track:[],
  };
  view.headline=view.stage==='entering'?'Buying: waiting for the order to fill':view.stage==='exiting'?`Selling: ${view.exitReason||'closing the position'}`:
    view.stage==='attention'?`Needs your attention: ${view.exitReason||'check Alpaca'}`:view.label;
  return view;
}

// Price since the entry with the entry, stop and target drawn as lines. Pure SVG string.
export function sparkline(view) {
  const points=(view?.track||[]).map(row=>({t:Date.parse(row?.t),p:num(row?.p)})).filter(row=>Number.isFinite(row.t)&&row.p!==null&&row.p>0);
  if(points.length<2)return '<p class="live-track-empty">The price line fills in as new quotes arrive (one point every 15 seconds).</p>';
  const levels=Object.entries(view.levels||{}).filter(([,v])=>v!==null&&v>0);
  const values=[...points.map(row=>row.p),...levels.map(([,v])=>v)];
  let low=Math.min(...values),high=Math.max(...values);
  const pad=(high-low)*0.08||high*0.001;low-=pad;high+=pad;
  const W=600,H=140,t0=points[0].t,t1=points.at(-1).t,span=Math.max(1,t1-t0);
  const x=t=>((t-t0)/span*W).toFixed(1),y=p=>((high-p)/(high-low)*H).toFixed(1);
  const line=points.map(row=>`${x(row.t)},${y(row.p)}`).join(' ');
  const guides=levels.map(([kind,v])=>`<line class="guide ${kind}" x1="0" x2="${W}" y1="${y(v)}" y2="${y(v)}" vector-effect="non-scaling-stroke"/>`).join('');
  const first=points[0],last=points.at(-1);
  return `<svg class="live-track" viewBox="0 0 ${W} ${H}" preserveAspectRatio="none" role="img" aria-label="${esc(`${view.symbol} price from ${timeET(new Date(first.t).toISOString())} to ${timeET(new Date(last.t).toISOString())}: ${price(first.p)} to ${price(last.p)}`)}">
    ${guides}<polyline class="price ${view.tone}" points="${line}" fill="none" vector-effect="non-scaling-stroke"/></svg>
    <p class="live-track-axis"><span>${esc(timeET(new Date(first.t).toISOString()))}</span><span>${esc(timeET(new Date(last.t).toISOString()))}</span></p>`;
}

function ladder(view) {
  if(view.progress===null)return '';
  const at=value=>`${(value*100).toFixed(1)}%`;
  return `<div class="live-ladder" role="img" aria-label="${esc(`Price is ${Math.round(view.progress*100)}% of the way from the stop ${view.stopPrice} to the target ${view.targetPrice}`)}">
    <div class="ladder-track">${view.entryAt!==null?`<i class="ladder-entry" style="left:${at(view.entryAt)}" title="Entry"></i>`:''}<i class="ladder-now ${view.tone}" style="left:${at(view.progress)}"></i></div>
    <div class="ladder-labels"><span class="stop">Stop ${esc(view.stopPrice)}</span>${view.entryAt!==null&&view.entryAt>0.2&&view.entryAt<0.8?`<span class="entry" style="left:${at(view.entryAt)}">Entry</span>`:''}<span class="target">Target ${esc(view.targetPrice)}</span></div></div>`;
}

export function liveTradeMarkup(live,options={}) {
  const view=liveTradeView(live,options);if(!view)return '';
  const plan=[];
  if(view.target)plan.push(`<li class="plan-target"><div><b>Target ${esc(view.target.price)}</b><span>${esc(view.target.result)} (${esc(view.target.percent)}) if reached${view.target.away?` · ${esc(view.target.away)}`:''}</span></div>
    <small>The app sells at market as soon as the ${esc(view.symbol)} bid reaches the target. It checks every 5 seconds.</small></li>`);
  if(view.stop)plan.push(`<li class="plan-stop"><div><b>Stop ${esc(view.stop.price)}</b><span>${esc(view.stop.result)} (${esc(view.stop.percent)}) if hit${view.stop.away?` · ${esc(view.stop.away)}`:''}</span></div>
    <small>A stop order at Alpaca sells automatically if the price falls to the stop. Status: <span class="${view.stop.working?'ok':'wait'}">${esc(view.stop.order)}</span>.</small></li>`);
  plan.push(`<li class="plan-close"><div><b>Market close${view.close?` · ${esc(view.close.at)}`:''}</b><span>${view.close?esc(view.close.left):'Next session'}</span></div>
    <small>If neither is hit, the app sells at market 5 minutes before the close. Socrates never holds overnight.</small></li>`);
  const signal=view.proxy&&view.signal?`<p class="help live-proxy">This is a QQQ <b>short</b> setup. The account cannot short, so the app bought PSQ, which rises when QQQ falls. QQQ now ${esc(price(view.signal.current))} · QQQ entry ${esc(price(view.signal.entry))} · QQQ stop ${esc(price(view.signal.stop))} · QQQ target ${esc(price(view.signal.target))}.</p>`:'';
  return `<article class="card live-trade" data-tone="${view.tone}" data-stage="${esc(view.stage)}" aria-labelledby="live-trade-title">
  <div class="live-head"><h2 id="live-trade-title"><span class="live-dot${view.stale?' stale':''}" aria-hidden="true"></span>Live trade · ${esc(view.symbol)}</h2><span class="pill">${esc(view.stageText)}</span><span class="live-updated${view.stale?' stale':''}">${esc(view.updated)}</span></div>
  <p class="live-headline">${esc(view.headline)}</p>
  <div class="live-numbers">
    <div class="live-pnl"><span>Up or down since the purchase</span><strong>${esc(view.pnl)}</strong><em>${esc(view.pnlDetail)}</em></div>
    <div><span>${esc(view.nowLabel)}</span><b>${esc(view.now)}</b></div>
    <div><span>Paid</span><b>${esc(view.cost)}</b></div>
    <div><span>Worth now</span><b>${esc(view.value)}</b></div>
  </div>
  <p class="live-bought">${esc(view.bought)}</p>
  ${ladder(view)}
  <div class="live-chart">${sparkline(view)}</div>
  <h3 class="live-plan-title">How it will sell</h3>
  <ul class="live-plan">${plan.join('')}</ul>
  ${signal}
  ${view.message?`<p class="help">${esc(view.message)}</p>`:''}
</article>`;
}

export function lastTradeMarkup(last,{now=Date.now()}={}) {
  last=obj(last);if(!last||!last.entry_price)return '';
  const gross=num(last.gross_pnl),result=gross===null?'Result not verified yet':signedMoney(gross);
  const toneName=gross===null?'flat':gross>0.004?'up':gross<-0.004?'down':'flat';
  const sold=last.exit_price?`sold at ${price(last.exit_price)}${last.completed_at?` · ${timeET(last.completed_at)}`:''}`:'sale price not confirmed';
  const day=dayET(last.completed_at||last.entered_at,now);
  return `<article class="card live-trade last-trade" data-tone="${toneName}" aria-labelledby="last-trade-title">
  <div class="live-head"><h2 id="last-trade-title">No open trade</h2><span class="pill">Last trade${day?` · ${esc(day)}`:''}</span></div>
  <p class="live-headline">${esc(text(last.label)||'Socrates trade')}: <strong class="last-result">${esc(result)}</strong>${last.percent?` (${esc(signedPercent(last.percent))})`:''}</p>
  <p class="live-bought">Bought ${esc(shares(last.quantity))} ${esc(text(last.symbol))} at ${esc(price(last.entry_price))}${last.entered_at?` · ${esc(timeET(last.entered_at))}`:''}, ${esc(sold)}.</p>
  ${text(last.exit_reason)?`<p class="help">Why it sold: ${esc(text(last.exit_reason))}.</p>`:''}
  ${gross!==null?'<p class="help">Gross result from Alpaca fills; fees are not included.</p>':''}
</article>`;
}

File: pivot/live_trade.py
"""Live view of the open Socrates trade (4.5.2): read-only arithmetic, no broker calls.

The executor records the same bid and ask it uses to decide a target exit. This
module combines that quote with the durable trade record, the Alpaca position and
the broker clock, and returns what the owner watches:
- the entry and the current price;
- the profit or loss in dollars, percent and R (multiples of the planned risk);
- the distance to the stop and to the target;
- the three ways the position will be sold.
Missing or stale evidence is labelled, never filled in. Nothing here can change
a trade or an order.
"""
from datetime import datetime, timezone
from decimal import Decimal, DecimalException, ROUND_HALF_UP

from .models import timestamp
from .performance import socrates_labels

# Mirrors execution.closing(): the session-close exit starts five minutes before the close.
CLOSE_EXIT_SECONDS = 300
# A mark older than this reads as delayed. The executor itself acts only on quotes under 15 seconds old.
FRESH_SECONDS = 20
POSITION_FALLBACK_SECONDS = 60
TRACK_POINTS = 240
STAGES = {'entering': 'Buying', 'open': 'Holding', 'exiting': 'Selling', 'attention': 'Needs your attention'}
STOP_ORDER_TEXT = {
    'new': 'working at Alpaca', 'accepted': 'accepted by Alpaca', 'pending_new': 'being accepted by Alpaca',
    'partially_filled': 'partly filled', 'filled': 'filled: the stop sold the shares',
    'canceled': 'canceled', 'expired': 'expired', 'rejected': 'rejected by Alpaca',
    'pending_cancel': 'being canceled', 'done_for_day': 'done for the day', 'suspended': 'suspended',
}
WORKING_STOP = {'new', 'accepted', 'pending_new', 'partially_filled'}


def _number(value):
    if value is None or isinstance(value, bool):
        return None
    try:
        number = Decimal(str(value))
    except (DecimalException, ValueError, TypeError):
        return None
    return number if number.is_finite() else None


def _money(value, places='0.01'):
    return str(value.quantize(Decimal(places), rounding=ROUND_HALF_UP)) if value is not None else None


def _at(value):
    try:
        at = timestamp(value)
    except (KeyError, TypeError, ValueError, OverflowError, AttributeError):
        return None
    return at if at.tzinfo is not None and at.utcoffset() is not None else None


def _ratio(part, whole, places='0.0001'):
    if part is None or whole is None or whole == 0:
        return None
    return str((part / whole).quantize(Decimal(places), rounding=ROUND_HALF_UP))


def _entry_fill(trade, position):
    seen = ((trade.get('ops') or {}).get('entry') or {}).get('last_seen') or {}
    price, qty = _number(seen.get('filled_avg_price')), _number(trade.get('filled_qty') or seen.get('filled_qty'))
    source = 'fill'
    if (price is None or price <= 0) and position:
        price, source = _number(position.get('avg_entry_price')), 'position'
    if (qty is None or qty <= 0) and position:
        qty = _number(position.get('qty'))
    at = _at(seen.get('filled_at')) or _at(trade.get('created_at'))
    if price is None or price <= 0:
        price, source = None, None
    if qty is not None and qty <= 0:
        qty = None
    return price, qty, at, source


def _current(trade, mark, position, now):
    """Sale-side price: the bid for a held purchase (what selling would get), the ask for a short."""
    long = trade.get('direction') != 'short'
    if isinstance(mark, dict) and mark.get('trade_id') == trade.get('id') and mark.get('symbol') == trade.get('symbol'):
        price = _number(mark.get('bid') if long else mark.get('ask'))
        at = _at(mark.get('quote_at')) or _at(mark.get('at'))
        if price is not None and price > 0 and at is not None:
            age = (now - at).total_seconds()
            # While the manager is not reading quotes (selling, or quotes paused) the
            # account worker's position price, at most 15 seconds old, is newer.
            if age <= POSITION_FALLBACK_SECONDS or not _number((position or {}).get('current_price')):
                return {'price': price, 'bid': _number(mark.get('bid')), 'ask': _number(mark.get('ask')),
                        'at': at, 'age': age, 'source': 'quote'}
    price = _number((position or {}).get('current_price'))
    if price is not None and price > 0:
        return {'price': price, 'bid': None, 'ask': None, 'at': None, 'age': None, 'source': 'position'}
    return None


def _stop_order(trade, orders):
    op = (trade.get('ops') or {}).get('stop')
    if not isinstance(op, dict):
        return {'status': None, 'text': 'not placed yet: the app places it right after the purchase fills', 'working': False}
    if op.get('state') == 'prepared':
        return {'status': 'preparing', 'text': 'being placed', 'working': False}
    client_id = (op.get('payload') or {}).get('client_order_id')
    live = next((o for o in orders or [] if isinstance(o, dict) and client_id and o.get('client_order_id') == client_id), None)
    status = (live or {}).get('status') or (op.get('last_seen') or {}).get('status')
    text = STOP_ORDER_TEXT.get(status, status.replace('_', ' ') if isinstance(status, str) else 'status not confirmed yet')
    return {'status': status if isinstance(status, str) else None, 'text': text, 'working': status in WORKING_STOP}


def _session_exit(clock, now):
    close = _at((clock or {}).get('next_close'))
    if close is None or not (clock or {}).get('is_open'):
        return None
    exit_at = close.timestamp() - CLOSE_EXIT_SECONDS
    return {'at': datetime.fromtimestamp(exit_at, timezone.utc).isoformat(), 'close_at': close.isoformat(),
            'seconds_left': max(0, round(exit_at - now.timestamp()))}


def _track(track, trade, entry_price):
    points = [row for row in (track or []) if isinstance(row, dict) and row.get('trade_id') == trade.get('id')]
    if len(points) > TRACK_POINTS:
        step = len(points) / TRACK_POINTS
        points = [points[int(index * step)] for index in range(TRACK_POINTS - 1)] + [points[-1]]
    long = trade.get('direction') != 'short'
    out = []
    for row in points:
        price = _number(row.get('bid') if long else row.get('ask'))
        at = _at(row.get('quote_at')) or _at(row.get('at'))
        if price is not None and price > 0 and at is not None:
            out.append({'t': at.isoformat(), 'p': str(price)})
    return out


def build_live_trade(trade, *, mark=None, track=None, positions=None, orders=None, clock=None,
                     signal_quote=None, message=None, now=None):
    """Owner-facing summary of the one Socrates trade the app is managing; None when there is none."""
    if not isinstance(trade, dict) or trade.get('stage') not in STAGES:
        return None
    now = now or datetime.now(timezone.utc)
    symbol = trade.get('symbol') if isinstance(trade.get('symbol'), str) else None
    long = trade.get('direction') != 'short'
    position = next((p for p in positions or [] if isinstance(p, dict) and p.get('symbol') == symbol), None)
    entry_price, qty, entered_at, entry_source = _entry_fill(trade, position)
    stop, target = _number(trade.get('stop')), _number(trade.get('target'))
    current = _current(trade, mark, position, now)
    price = current['price'] if current else None
    labels = socrates_labels(trade)
    result = {
        'trade_id': trade.get('id') if isinstance(trade.get('id'), str) else None,
        'stage': trade['stage'], 'stage_text': STAGES[trade['stage']],
        'symbol': symbol, 'direction': 'short' if not long else 'long',
        'label': labels['label'] or (f'Socrates {symbol}' if symbol else 'Socrates trade'),
        'proxy': labels['proxy'] == 'inverse_etf',
        'signal_direction': labels['signal_direction'],
        'amount': _money(_number(trade.get('amount'))),
        'entry': {'price': str(entry_price) if entry_price else None, 'qty': str(qty) if qty else None,
                  'cost': _money(entry_price * qty) if entry_price and qty else None,
                  'at': entered_at.isoformat() if entered_at else None, 'source': entry_source},
        'current': None, 'pnl': None, 'stop': None, 'target': None, 'progress': None,
        'session_exit': _session_exit(clock, now), 'signal': None, 'exit': None, 'track': [],
        'message': message if isinstance(message, str) else None,
        'as_of': now.isoformat(),
    }
    if current:
        result['current'] = {'price': str(price), 'bid': str(current['bid']) if current['bid'] else None,
                             'ask': str(current['ask']) if current['ask'] else None,
                             'at': current['at'].isoformat() if current['at'] else None,
                             'source': current['source'],
                             'age_seconds': round(current['age']) if current['age'] is not None else None,
                             'fresh': current['source'] == 'quote' and current['age'] is not None and current['age'] <= FRESH_SECONDS}
    if entry_price and price:
        move = (price - entry_price) if long else (entry_price - price)
        risk = (entry_price - stop) if (stop and long) else (stop - entry_price) if stop else None
        result['pnl'] = {'dollars': _money(move * qty) if qty else None,
                         'per_share': str(move.quantize(Decimal('0.0001'), rounding=ROUND_HALF_UP)),
                         'percent': _ratio(move * 100, entry_price, '0.01'),
                         'r': _ratio(move, risk, '0.01') if risk and risk > 0 else None}
    for name, level in (('stop', stop), ('target', target)):
        if level is None or level <= 0:
            continue
        row = {'price': str(level)}
        if entry_price and qty:
            change = (level - entry_price) if long else (entry_price - level)
            row['result_if_hit'] = _money(change * qty)
            row['result_percent'] = _ratio(change * 100, entry_price, '0.01')
        if price:
            # Positive: room left before the level is reached.
            gap = (price - level) if (name == 'stop') == long else (level - price)
            row['distance'] = str(gap.quantize(Decimal('0.0001'), rounding=ROUND_HALF_UP))
            row['distance_percent'] = _ratio(gap * 100, price, '0.01')
        result[name] = row
    if result['target'] is not None:
        from .strategy import KEY_LEVEL_NAMES
        source = trade.get('target_source') or (trade.get('signal_geometry') or {}).get('target_source')
        result['target']['source'] = KEY_LEVEL_NAMES.get(source, source) if isinstance(source, str) else None
    if result['stop'] is not None:
        result['stop']['order'] = _stop_order(trade, orders)
    if stop and target and price and stop != target:
        span = (target - stop) if long else (stop - target)
        where = ((price - stop) if long else (stop - price)) / span
        result['progress'] = str(max(Decimal(0), min(Decimal(1), where)).quantize(Decimal('0.001')))
    geometry = trade.get('signal_geometry') if isinstance(trade.get('signal_geometry'), dict) else None
    if result['proxy'] and geometry:
        qqq = _number((signal_quote or {}).get('bp'))
        result['signal'] = {'symbol': 'QQQ', 'direction': 'short', 'entry': geometry.get('entry'),
                            'stop': geometry.get('stop'), 'target': geometry.get('target'),
                            'current': str(qqq) if qqq and qqq > 0 else None,
                            'at': (signal_quote or {}).get('t') if qqq else None}
    if trade['stage'] == 'exiting':
        pending = trade.get('exit_pending') if isinstance(trade.get('exit_pending'), dict) else {}
        reason = trade.get('exit_reason') or trade.get('reason')
        result['exit'] = {'reason': reason if isinstance(reason, str) else 'Closing the position',
                          'started_at': pending.get('first_observed_at')}
    elif trade['stage'] == 'attention':
        result['exit'] = {'reason': trade.get('reason') if isinstance(trade.get('reason'), str) else 'Needs review in Alpaca',
                          'started_at': None}
    result['track'] = _track(track, trade, entry_price)
    return result


def build_last_trade(trade):
    """The most recent finished Socrates trade that bought shares: prices, result and why it sold.

    The gross result comes from performance.trade_result, which verifies every
    fill. The average prices shown next to it are read from the same broker fill
    records and are labelled unverified when that check fails.
    """
    if not isinstance(trade, dict) or trade.get('stage') != 'finished':
        return None
    from .performance import trade_result
    ops = trade.get('ops') if isinstance(trade.get('ops'), dict) else {}
    seen = (ops.get('entry') or {}).get('last_seen') or {}
    entry_qty, entry_price = _number(seen.get('filled_qty')), _number(seen.get('filled_avg_price'))
    if not entry_qty or entry_qty <= 0 or not entry_price or entry_price <= 0:
        return None
    exit_qty, exit_value, stop_filled = Decimal(0), Decimal(0), False
    for name, op in ops.items():
        if name == 'entry' or not isinstance(op, dict):
            continue
        fill = op.get('last_seen') or {}
        qty, price = _number(fill.get('filled_qty')), _number(fill.get('filled_avg_price'))
        if qty and qty > 0 and price and price > 0:
            exit_qty += qty
            exit_value += qty * price
            stop_filled = stop_filled or name == 'stop'
    result = trade_result(trade)
    reason = trade.get('exit_reason') if isinstance(trade.get('exit_reason'), str) else None
    if stop_filled and not reason:
        reason = 'The broker stop order sold the shares'
    exit_price = exit_value / exit_qty if exit_qty > 0 else None
    long = trade.get('direction') != 'short'
    percent = None
    if exit_price:
        move = (exit_price - entry_price) if long else (entry_price - exit_price)
        percent = _ratio(move * 100, entry_price, '0.01')
    entered = _at(seen.get('filled_at')) or _at(trade.get('created_at'))
    return {'trade_id': trade.get('id') if isinstance(trade.get('id'), str) else None,
            'label': result.get('label') or (f'Socrates {trade.get("symbol")}' if trade.get('symbol') else 'Socrates trade'),
            'symbol': result.get('symbol'), 'quantity': str(entry_qty),
            'entry_price': str(entry_price), 'entered_at': entered.isoformat() if entered else None,
            'exit_price': str(exit_price.quantize(Decimal('0.0001'), rounding=ROUND_HALF_UP)) if exit_price else None,
            'completed_at': result.get('completed_at'),
            'gross_pnl': _money(_number(result.get('gross_pnl'))) if result.get('status') == 'verified_gross' else None,
            'percent': percent, 'status': result.get('status'), 'exit_reason': reason,
            'stop': trade.get('stop'), 'target': trade.get('target')}
